### PR TITLE
[SPARK-16963] [STREAMING] [SQL] Changes to Source trait and related implementation classes

### PR DESCRIPTION
**Pull request by frreiss**
## What changes were proposed in this pull request?

This PR contains changes to the Source trait such that the scheduler can notify data sources when it is safe to discard buffered data. Summary of changes:
- Added a method `commit(end: Offset)` that tells the Source that is OK to discard all offsets up `end`, inclusive.
- Changed the semantics of a `None` value for the `getBatch` method to mean "from the very beginning of the stream"; as opposed to "all data present in the Source's buffer".
- Added notes that the upper layers of the system will never call `getBatch` with a start value less than the last value passed to `commit`.
- Added a `lastCommittedOffset` method to allow the scheduler to query the status of each Source on restart. This addition is not strictly necessary, but it seemed like a good idea -- Sources will be maintaining their own persistent state, and there may be bugs in the checkpointing code.
- The scheduler in `StreamExecution.scala` now calls `commit` on its stream sources after marking each batch as complete in its checkpoint.
- `MemoryStream` now cleans committed batches out of its internal buffer.
- `TextSocketSource` now cleans committed batches from its internal buffer.
## How was this patch tested?

Existing regression tests already exercise the new code.

### CODE REVIEW

**Comment by frreiss** [comment]
These changes are now ready for review. The contents of this PR pass regression tests on my machines. Can one of the committers please start a Jenkins build?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetadataLog.scala
Comment (wangmiao1981) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetadataLog.scala]: Extra blank line

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala
Comment (wangmiao1981) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Extra blank line

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/socket.scala
Comment (wangmiao1981) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/socket.scala]: Extra blank line

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala
Comment (wangmiao1981) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Extra blank line

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetadataLog.scala
Comment (frreiss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MetadataLog.scala]: Fixed in my local copy.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala
Comment (frreiss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Fixed in my local copy.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/socket.scala
Comment (frreiss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/socket.scala]: Fixed in my local copy.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala
Comment (frreiss) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: This line no longer exists after merging the changes from [https://github.com/apache/spark/pull/14728] and replacing my former implementation of GC for FileStreamSource with the one in that merged PR.

**Comment by frreiss** [comment]
@rxin and @marmbrus, would it be possible to get this PR reviewed soon? I can split it into smaller chunks if that would make things easier; I just need to know.

**Comment by frreiss** [comment]
@ScrapCodes, would you mind triggering a build of this PR?

**Comment by ScrapCodes** [comment]
ok to test

**Comment by ScrapCodes** [comment]
retest this please

**Comment by ScrapCodes** [comment]
I have tested the PR with my MQTT connector. Looks like I do not have sufficient privilege to command jenkins.

**Comment by vanzin** [comment]
ok to test

**Comment by zsxwing** [comment]
> Added a lastCommittedOffset method to allow the scheduler to query the status of each Source on restart. This addition is not strictly necessary, but it seemed like a good idea -- Sources will be maintaining their own persistent state, and there may be bugs in the checkpointing code.

I'm not sure if we add this API. This one forces Source should store `lastCommittedOffset` in some reliable storage, otherwise, Source will always return None when recovering from failure.

I suggest we add it later when we really need to use this API.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Source.scala
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Source.scala]: nit: not need `<p>` in scaladoc

**Comment by zsxwing** [comment]
Jenkins, add to whitelist.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: I agree with the first sentance.  I'm not sure I understand the second.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Source.scala
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Source.scala]: I'd agree to leave this out if its only used to output a warning message.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Source.scala]: Also, nit: no `()`

**Review by marmbrus — COMMENTED** [review]
Overall looks pretty good.  Thanks for working on this!

**Review by marmbrus — COMMENTED** [review]
It would be nice to merge this.  Any response on the open comments?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Source.scala
Comment (frreiss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Source.scala]: Fixed in my local copy.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Source.scala
Comment (frreiss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Source.scala]: Dropped this method from my local copy.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala
Comment (frreiss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: What I meant to say is that two accesses to this field from a thread other than the scheduler thread may return different different values. The functions that access `committedOffsets` from other threads either should access the field only once (`toDebugString`); or they should make a copy (as in the code that consumes `StreamingQueryStatus` objects). Updated the comment in my local copy to something that is hopefully more clear: "Other threads should make a shallow copy if they are going to access this field more than once, since the field's value may change at any time."

**Comment by frreiss** [comment]
Sorry, I missed the last few email notifications about this PR. I've merged with the head version and made updates to address the most recent round of review comments. Currently running regression tests locally.

**Comment by frreiss** [comment]
All my changes are in now, and regression tests pass. As far as I can see, all the review comments have been addressed at this point.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Can we just commit `currentBatchId - 1`? In which case, it will fail?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: nit: occured -> occurred

**Review by zsxwing — CHANGES_REQUESTED** [review]
Looks pretty good. Just one question about which offset should be committed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Source.scala
Comment (frreiss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Source.scala]: Fixed in my local copy.

**Comment by zsxwing** [comment]
@frreiss could you update the PR? Thanks!

**Comment by frreiss** [comment]
I've been running tests since this morning; should have updates in soon.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala
Comment (frreiss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Fixed in my local copy

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala
Comment (frreiss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Yes, `currentBatchId - 1` does work here, and I've updated my local copy.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: nit: this can be `offsetLog.purge(currentBatchId)`, it's exclusive, then you can revert changes to StreamingQuerySuite.

**Review by zsxwing — COMMENTED** [review]
LGTM except one nit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala
Comment (frreiss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: I can move this change to another JIRA if you'd like, but we really should change `currentBatchId` to `currentBatchId - 1` at some point. The call to `offsetLog.purge(currentBatchId)`, which I introduced in my PR for SPARK-17513, contains a subtle bug. The recovery logic in `populateStartOffsets()` reads the last and second-to-last entries in `offsetLog`. `populateStartOffsets()` uses those entries to populate `availableOffsets` and `committedOffsets`, respectively. Calling `offsetLog.purge(currentBatchId)` at line 350/366 results in the `offsetLog` being truncated to one entry, which in turn results in `committedOffsets` being left empty on recovery, which in turn causes the first call to `getBatch()` for any source to have `None` as its first argument. Sources that do not prune buffered data in their `commit()` methods will return a previously committed data in response to such a `getBatch()` call.

**Comment by zsxwing** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: I see. Thanks for your clarifying.

**Review by zsxwing — APPROVED** [review]
LGTM pending tests

**Comment by zsxwing** [comment]
@frreiss you need to reset `lastOffsetCommitted` in `MemoryStream.reset`. That's why the test fails.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala]: nit: 
```
end match {
  case newOffset: LongOffset =>
     ...
  case _ => sys.error(...)
}
```

**Comment by brkyvz** [comment]
LGTM as well!

**Comment by zsxwing** [comment]
@frreiss any update?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala
Comment (frreiss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/memory.scala]: Corrected in my local copy.

**Comment by frreiss** [comment]
Updated the branch and addressed new review comments. Looks like my last push missed a one-line change to memory.scala. Tests are running now.

**Comment by zsxwing** [comment]
LGMT. Merging to master and 2.0. Thanks!

**Pull request closed without merging**
